### DISCUSSION TITLE
Dynamic cron schedules

### DISCUSSION THREAD

**Discussion by abrunner94** [bensheldon/good_job#1237] [Q&A] [answered]
👋 I'm looking for a programmatic way to add cron schedules without defining them in my `application.rb` file.

For example, I would like to convert the following predefined schedule into plain Ruby.

`application.rb`
```ruby
    config.good_job.cron = {
      test_task: {
        cron: "* * * * *",
        class: "TestJob",
      }
    }
```

What I tried:
- Looked at `GoodJob::CronEntry` and `GoodJob::CronManager`, however I'm also unsure about whether these should/could indeed be used to accomplish this.

Does someone have an example if this is currently possible?

**Comment by bensheldon** [comment]
What do you mean by "plain ruby"? The configuration is a Ruby hash and you can use procs/lambdas to set values dynamically.

These values are tricky because the configuration must be set up when the application boots. 

What is feature in your application you're trying to do?

**Reply by bensheldon** [reply]
oh, maybe... Are you trying to have a user-facing feature in which users can define their own schedules?

**Reply by abrunner94** [reply]
Exactly! 

Elixir example using the Oban package:

https://hexdocs.pm/oban/2.11.0/dynamic_cron.html

```elixir
DynamicCron.insert([
  {"0 0 * * *", MyApp.GenericWorker},
  {"* * * * *", MyApp.ClientWorker, name: "client-1", args: %{client_id: 1}},
  {"* * * * *", MyApp.ClientWorker, name: "client-2", args: %{client_id: 2}},
  {"* * * * *", MyApp.ClientWorker, name: "client-3", args: %{client_id: 3}}
])

```

**Reply by bensheldon** [reply, marked as answer]
Unfortunately GoodJob doesn't have anything like that, and I don't think it's a good fit for GoodJob. GoodJob's cron all lives together in memory in each process, so each cron entry uses a little bit of memory and uses a little bit of CPU when it runs; it's not intended to be unbounded or dynamic

I'd imagine if you wanted something like this, that you'd build your own scheduler-like-behavior as part of your application: Run a job every minute, query for records that would be expected to run, run them, and then mark them as run until their next tick (e.g. just keep bumping their last_run_at and next_run_at)

**Reply by abrunner94** [reply]
> Run a job every minute, query for records that would be expected to run, run them, and then mark them as run until their next tick (e.g. just keep bumping their last_run_at and next_run_at)

I was just thinking the same - I'll go with that. Thank you!
